### PR TITLE
api - health - kube conventions

### PR DESCRIPTION
**Pull request by brandicorel**

### CODE REVIEW

File: src/AspNetCore/Configurator/MonitoringConfigurator.HealthCheck.cs [hunk 1]
```diff
@@ -21,12 +22,44 @@ public static IServiceCollection AddHealthChecks(this IServiceCollection service
...
+                    options.MapHealthChecks("/healthz/live",    _ => false);
+                    options.MapHealthChecks("/healthz/startup", "startup");
+                    options.MapHealthChecks("/healthz/ready",   "ready");
+                });
+        }
+        
         public static IEndpointConventionBuilder MapHealthChecks(this IEndpointRouteBuilder options)
+        {
+            return options.MapHealthChecks("/health");
+        }
```
Comment (KevinRecuerda) [inline src/AspNetCore/Configurator/MonitoringConfigurator.HealthCheck.cs:46]: maybe we can trash this

File: src/AspNetCore/Configurator/MonitoringConfigurator.HealthCheck.cs [hunk 2]
```diff
@@ -21,12 +22,44 @@ public static IServiceCollection AddHealthChecks(this IServiceCollection service
...
+
+        public static IEndpointConventionBuilder MapHealthChecks(this IEndpointRouteBuilder options, string pattern, params string[] tags)
+        {
+            return options.MapHealthChecks(pattern, check => !tags.Any() || tags.Any(tag => check.Tags.Contains(tag)));
```
Comment (KevinRecuerda) [inline src/AspNetCore/Configurator/MonitoringConfigurator.HealthCheck.cs:50]: could use `ContainsAny`

File: src/AspNetCore/Health/HealthStartupService.cs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.Extensions.Hosting;
+
+    public interface IHealthStartupService
+    {
+        public bool HasCompleted { get; set; }
+    }
+
+    public abstract class BaseHealthStartupService : BackgroundService, IHealthStartupService
```
Comment (KevinRecuerda) [inline src/AspNetCore/Health/HealthStartupService.cs:12]: should we rename `BaseHealthStartupService` &rarr; `StartupBackgroundService` ?

File: src/AspNetCore/Health/HealthStartupService.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+        public bool HasCompleted { get; set; }
+
+        protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+        {
+            await this.InnerExecute();
+
+            this.HasCompleted = true;
+        }
+
+        public abstract Task InnerExecute();
```
Comment (KevinRecuerda) [inline src/AspNetCore/Health/HealthStartupService.cs:23]: should pass cancellationToken

File: src/AspNetCore/Health/StartupHealthService.cs [hunk 5]
```diff
@@ -0,0 +1,29 @@
+﻿namespace RecShark.AspNetCore.Health
+{
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.Extensions.Diagnostics.HealthChecks;
+
+    public class StartupHealthService : IHealthCheck
```
Comment (KevinRecuerda) [inline src/AspNetCore/Health/StartupHealthService.cs:9]: could rename `StartupHealthService` &rarr; `StartupHealthChecker`

File: src/AspNetCore/Configurator/MonitoringConfigurator.HealthCheck.cs [hunk 6]
```diff
@@ -21,12 +22,44 @@ public static IServiceCollection AddHealthChecks(this IServiceCollection service
             return services;
         }
 
+        public static void AddStartupHealthService<TStartupService>(this IServiceCollection services)
+            where TStartupService : BaseHealthStartupService
+        {
+            services.AddSingleton<IHealthStartupService, TStartupService>();
+            services.AddHostedService(sp => sp.GetService<IHealthStartupService>() as TStartupService);
```
Comment (KevinRecuerda) [inline src/AspNetCore/Configurator/MonitoringConfigurator.HealthCheck.cs:29]: can't register multiple services.
could do that:
```csharp
services.AddSingleton<TStartupService>();
services.Add<IHealthStartupService>((sp => sp.GetService<TStartupService>());
services.AddHostedService(sp => sp.GetService<TStartupService>());
```

File: src/AspNetCore/Configurator/MonitoringConfigurator.HealthCheck.cs [hunk 7]
```diff
@@ -21,12 +22,44 @@ public static IServiceCollection AddHealthChecks(this IServiceCollection service
             return services;
         }
 
+        public static void AddStartupHealthService<TStartupService>(this IServiceCollection services)
```
Comment (KevinRecuerda) [inline src/AspNetCore/Configurator/MonitoringConfigurator.HealthCheck.cs:25]: should we rename to `StartupHostedService` ?

File: src/AspNetCore/Health/StartupHealthService.cs [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+        {
+            this.Services = services.ToList();
+        }
+
+        private List<IHealthStartupService> Services { get; }
+
+        public Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, CancellationToken cancellationToken = default)
+        {
+            if (!this.Services.Any())
+                return Task.FromResult(HealthCheckResult.Healthy("No startup task defined."));
```
Comment (KevinRecuerda) [inline src/AspNetCore/Health/StartupHealthService.cs:21]: maybe we could remove this part.

File: src/AspNetCore/Configurator/MonitoringConfigurator.HealthCheck.cs [hunk 9]
```diff
@@ -21,12 +22,44 @@ public static IServiceCollection AddHealthChecks(this IServiceCollection service
...
+            services.AddHostedService(sp => sp.GetService<IHealthStartupService>() as TStartupService);
+        }
+
+        public static IApplicationBuilder MapHealthChecks(this IApplicationBuilder app)
+        {
+            return app.UseEndpoints(
+                options =>
+                {
+                    options.MapHealthChecks("/healthz/live",    _ => false);
+                    options.MapHealthChecks("/healthz/startup", "startup");
```
Comment (KevinRecuerda) [inline src/AspNetCore/Configurator/MonitoringConfigurator.HealthCheck.cs:38]: i think we should add `StartupHealthChecker` by default

**Review by KevinRecuerda — CHANGES_REQUESTED** [review]

File: src/AspNetCore/Configurator/MonitoringConfigurator.HealthCheck.cs (hunk 1 above)
Comment (brandicorel) [reply src/AspNetCore/Configurator/MonitoringConfigurator.HealthCheck.cs:46]: I've left it for retro-compatibility but I agree we can remove it.

File: src/AspNetCore/Health/HealthStartupService.cs (hunk 3 above)
Comment (brandicorel) [reply src/AspNetCore/Health/HealthStartupService.cs:12]: it's an abstract class, I though usage was to prefix abstract class by "Base"

File: src/AspNetCore/Configurator/MonitoringConfigurator.HealthCheck.cs [hunk 10]
```diff
@@ -12,21 +12,25 @@ namespace RecShark.AspNetCore.Configurator
...
+                         .AddHealthChecks()
+                         .AddCheck<StartupHealthChecker>("initialize", tags: new[] {"startup"});
             healthCheckBuilder?.Invoke(builder);
             return services;
         }
 
-        public static void AddStartupHealthService<TStartupService>(this IServiceCollection services)
-            where TStartupService : BaseHealthStartupService
+        public static void AddStartupHostedService<TStartupHostedService>(this IServiceCollection services)
+            where TStartupHostedService : class, IStartupHostedService
```
Comment (KevinRecuerda) [inline src/AspNetCore/Configurator/MonitoringConfigurator.HealthCheck.cs:29]: could rename `TStartupHostedService` to `T`
(not easy to distinguish `TStartupHostedService` and `IStartupHostedService`)

File: src/AspNetCore/Configurator/MonitoringConfigurator.HealthCheck.cs [hunk 11]
```diff
@@ -12,21 +12,25 @@ namespace RecShark.AspNetCore.Configurator
...
+    using RecShark.Extensions;
 
     public static class HealthCheckConfigurator
     {
         public static IServiceCollection AddHealthChecks(this IServiceCollection services, Action<IHealthChecksBuilder> healthCheckBuilder)
         {
-            var builder = services.AddHealthChecks();
+            var builder = services
+                         .AddHealthChecks()
+                         .AddCheck<StartupHealthChecker>("initialize", tags: new[] {"startup"});
```
Comment (KevinRecuerda) [inline src/AspNetCore/Configurator/MonitoringConfigurator.HealthCheck.cs:23]: should we name this check `startup` ?

File: sample/AspNetCore.Sample/Startup.cs [hunk 12]
```diff
@@ -53,12 +53,8 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env, IHostApp
...
                 // .UseSecurity()
                .UseOA3Swagger()
-               .UseEndpoints(
-                    endpoints =>
-                    {
-                        endpoints.MapHealthChecks();
-                        endpoints.MapControllers();
-                    });
+               .MapHealthChecks()
+               .UseEndpoints(endpoints => { endpoints.MapControllers(); });
```
Comment (KevinRecuerda) [inline sample/AspNetCore.Sample/Startup.cs:57]: old way seems more appropriate
(extension only on `endpoints`: `endpoints.MapConventionalHealthChecks()` ?)

File: src/AspNetCore/Health/StartupHealthChecker.cs [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+        {
+            this.Services = services.ToList();
+        }
+
+        private List<IStartupHostedService> Services { get; }
+
+        public Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, CancellationToken cancellationToken = default)
+        {
+            if (this.Services.Any(s => !s.HasCompleted))
+                return Task.FromResult(HealthCheckResult.Unhealthy("The startup tasks are still running."));
```
Comment (KevinRecuerda) [inline src/AspNetCore/Health/StartupHealthChecker.cs:21]: should we list running tasks ?

**Review by KevinRecuerda — CHANGES_REQUESTED** [review]

**Review by KevinRecuerda — APPROVED** [review]

**Pull request merged**